adjust TAF store directives for mom_fluxform.F and mom_vecinv.F, solve_tri/pentadiagonal.F, and calc_phi_hyd.F

mjlosch opened the pull request:
## What changes does this PR introduce?
cleanup and suggestion

## What is the current behaviour? 
- part of the routine bodies of `mom_fluxform.F` and `mom_vecinv.F` are recomputed in AD mode;
`AUTODIFF_ALLOW_VISCFACADJ` related code only in `mom_vecinv.F`
- recomputation warnings in `verification/global_oce_ biogeo_bling` because of `solve_tri/pentadiagnonal.F` with `SOLVE_DIAGONAL_KINNER` defined
- lots of warnings that make it difficult to read taf_ad.log

## What is the new behaviour 
- less of the routine bodies are recomputed in AD mode at the cost of increased memory overhead
- `AUTODIFF_ALLOW_VISCFACADJ` related code added to `mom_fluxform.F` to be consistent with `mom_vecinv.F`
- in general, storing strategies are more consistent between `mom_fluxform.F` and `mom_vecinv.F`
- avoid recomputing routine body of model/src/calc_phi_hyd.F by more store directives
- no more recomputation warnings in `verification/global_oce_ biogeo_bling`
- remove conflict warnings and some unused tape and "storing not necessary" warnings
- remove unnecessary checkpointing code from `pkg/smooth_diff2/3d.F`

## Does this PR introduce a breaking change? 
no, but storing `cDrag` in `mom_vecinv.F` (part of [d1a0161](https://github.com/MITgcm/MITgcm/pull/723/commits/d1a0161a0621567d537b5cf172c45b6c04cfc49e)) changes results at the working precision level: very small change in `obcs_ctrl` and small changes in the admonitor output in `verification_other/global_oce_cs32` and `verification_other/global_oce_llc90`, too small to modify the tested gradient.

## Other information:
This is a suggestion. Some stores can be debated and I would like to hear other opinions.

## Suggested addition to `tag-index`
- pkg/mom_fluxflux/vecinv: rearrange and add storing of variables to reduce number of (unwarned) recomputations.
- model/src/solve_tri/pentadiagonal.F: store directives to avoid recomputation warnings with SOLVE_DIAGONAL_KINNER defined
- compile routine bodies gad_som*.F only if GAD_ALLOW_TS_SOM_ADV or PTRACERS_ALLOW_DYN_STATE are defined
- model/src/calc_phi_hyd.F: avoid recomputing routine body by more store directives
- remove and refine many unnecessary store directives to reduce the number of TAF warnings
- remove unnecessary checkpointing code from pkg/smooth_diff2/3d.F

mjlosch commented:
@owang01 I am still not quite done with my cleanup of store directives, but the most important changes have been made. What's left ist some stuff in the bio-packages, where I have a hard time deciding what's important and shouldn't be touched (maybe @jahn, @seamanticscience, @mmazloff have an idea).

In the smooth package, which is only tested in `verification_other/global_oce_cs32/llc90`, there is some `TAMC_CHECKPOINTING` code in `smooth_diff2/3.F`. I think that this code can be removed entirely and there is no need to store anything in those routines (TAF ignores all store directives and only complains about some of them not being necessary), but I am a little reluctant to push this code, because I don't know what the implications will be for other ECCO setups (@owang01 , @antnguyen13). If you want to try this, I can push one commit, which we could revert if it doesn't work for ECCO (as any of the others, as they are all independent).

owang01 reviewed (commented):
@mjlosch I tested this branch on ECCO V4r5 for 3-month runs and the differences of adjoint gradients are very small (<1e-6) between this branch and the master. The caveat of course is that not all of the changes are actually tested, as V4r5 does not use some of the packages that this PR modifies. In terms of code performance, the run using this branch took similar wall time as that using the master. This probably does not mean much as Pleiades uses Lustre file system that makes code performance fluctuate a lot. As you removed many unnecessary stores, it is reasonably to believe that this branch will run faster.

mjlosch commented on model/src/solve_pentadiagonal.F:
```diff
@@ -252,7 +272,7 @@ SUBROUTINE SOLVE_PENTADIAGONAL(
      &      -(b5d(i,j,k)-a5d(i,j,k)*d5d_prime(k-2))*y5d_prime(k-1)
           ENDIF
-
+x
```
Oops, who did that even compile? I'll fix that.

mjlosch commented on pkg/generic_advdiff/gad_som_adv_x.F:
```diff
@@ -236,7 +237,7 @@ SUBROUTINE GAD_SOM_ADV_X(
 
 #ifdef ALLOW_AUTODIFF_TAMC
-CADJ STORE sm_o,sm_v,sm_x,sm_xx,sm_xy,sm_xz,sm_y,sm_yy,sm_yz,sm_z,sm_zz
+CADJ STORE sm_o,sm_v,sm_x,sm_y,sm_z,sm_xx,sm_xy,sm_xz
```
When you add these back, then you get "storing  sm_yy, sm_yz, sm_zz not necessary". I guess this is because the corresponding updates are linear, but for the others it's not? Do you get more recomputations or even warnings with this change? If so, which setup should I use to reproduce that?

owang01 replied:
That makes sense. I did not get more recomputations or warnings without storing sm_yy, sm_yz, and sm_zz.

mjlosch commented:
@owang01 thanks!

>  I tested this branch on ECCO V4r5 for 3-month runs and the differences of adjoint gradients are very small (<1e-6) between this branch and the master.

As the storing strategy is now a little different, we can expect small differences (especially since you define SOLVE_DIAGONAL_KINNER, which used to do recomputations, with warnings, which should now be gone, right?). 

> In terms of code performance, the run using this branch took similar wall time as that using the master. This probably does not mean much as Pleiades uses Lustre file system that makes code performance fluctuate a lot. 

The changes (for SOLVE_DIAGONAL_KINNER) may make things a slightlye faster, because some recomputations are avoided, but not very much, so that's also consistent with your test. 

> As you removed many unnecessary store directives, it is reasonably to believe that this branch will run faster.

I don't think (I hope!!!) that my cleanup of removing many unnecessary store directives changes anything in the AD code, just cleans up the taf_ad.log file a lot.

antnguyen13 commented on pkg/salt_plume/salt_plume_calc_depth.F:
```diff
@@ -76,6 +76,10 @@ SUBROUTINE SALT_PLUME_CALC_DEPTH(
 
+#ifdef ALLOW_AUTODIFF_TAMC
+C     suppress a warning
+CADJ INCOMPLETE SaltPlumeDepth(:,:,bi,bi), SPIND
```
this should be SaltPlumeDepth(:,:,bi,bj) <-- bj , not twice bi

mjlosch replied:
oh, wow, thanks for catching this! When did this get in (ie. when did I introduce that)?

mjlosch commented:
@antnguyen13 
about the `EXCLUDE_PCELL_MIX_CODE`: as far as I remember this flag was put there explicitly to not show the corresponding code to TAF. Not sure if we ***want*** to make it work with TAF because no-one has ever tested if the corresponding code produces stable AD-runs. Unless you set some of the runtime parameters explicitly the code is not used anyway (and I don't see these parameters https://github.com/MITgcm-contrib/llc_hires/tree/master/llc_90/ecco_v4r4/input/data). 
I think I already fixed it for `calc_3d_diffusivities.F` (for some reason), so I am surprised that it still produces these warnings. How do I reproduce the problem?

I think we should just not use `CHECK_SALINITY_FOR_NEGATIVE_VALUES` in AD runs. For the nonlinear free surface with real freshwater flux it does not make too much sense anyway.

mjlosch commented:
@antnguyen13 please try the latest commits with your code to see if it helps (with the `EXCLUDE_PCELL_MIX_CODE`). Quite some additional stores are required, so I would only use this if really necessary

jm-c commented:
@mjlosch there was 2 conflicting files when I merged master in: for `packages_init_variables.F` it was easy to fix but for `load_fields_driver.F` I ended up adding back all the included header files you removed earlier since many more have been added for Tapenade. Doing the other way (removing them) would have revert some of the changes from PR #685, and this is not what a merge commit is supposed to do. Anyway, this will need to be sorted out (and making sure all the Tapenade tests are OK).

owang01 commented on model/src/calc_3d_diffusivity.F:
```diff
@@ -364,6 +365,9 @@ SUBROUTINE CALC_3D_DIFFUSIVITY(
            ENDDO
           ENDDO
           IF ( mixSurf.EQ.2 ) THEN
+# ifdef ALLOW_AUTODIFF_TAMC
+CADJ STORE tmpFac          = loctape_3d_diff, key = k
+# endif
```
Shouldn't lines 368-370 be moved to above line 367 `IF ( mixSurf.EQ.2 ) THEN`? Or do we really need the store derivative here? The recomputation seems not expensive.

mjlosch replied:
yes, it makes more sense to move it out of the if-statement. To recompute `tmpFac` here is not so expensive, but this version of `tmpFac` is recomputed two times,  and I just wanted to avoid going through the if-statement within a loop multiple times. Does that make sense?

owang01 replied:
Thanks. That makes sense.

owang01 reviewed (commented):
@mjlosch With the latest commit https://github.com/MITgcm/MITgcm/pull/723/commits/5ac41c9fb611cb46898b49482a24c1db24270406, I was able to get identical adjoint gradients with those using commit https://github.com/MITgcm/MITgcm/pull/723/commits/b94cc86574c4dafbcdb7bc4266dcf13fbf0db3ab, which is good. Furthermore, the TAF recomputation warnings are mostly gone in https://github.com/MITgcm/MITgcm/pull/723/commits/5ac41c9fb611cb46898b49482a24c1db24270406. 

Since several pieces of the TAMC_CHECKPOINTING code in smooth_diff2/3.F have been removed, should we mention it in the PR summary?

jm-c commented:
Only did a quick look at these changes:

1. why adding a local version of `GAD_OPTIONS.h` in `tutorial_advection_in_gyre/code` ? SOM advection is not used there but it's already compiled (and run) in other exp. (e.g., all 3 advect_[cs/xy/xz] + in AD exp. `tutorial_tracer_adjsens` ).
2. I run testreport, no changes to FWD test but little changes to `obcs_ctrl` and AD cs32 test from `verification_other`. Will run the llc90 AD test later (it takes time) but would be nice to know where these changes come from.

mjlosch commented:
Thanks for looking at this (again)!

@jm-c 
1. with commit [1ee65c6](https://github.com/MITgcm/MITgcm/pull/723/commits/1ee65c635996bb6b269ba41ce8704482f89c19fa), using SOM advection with `ptracers` requires defining `GAD_ALLOW_TS_SOM_ADV` (see "Other information" in PR description). I implemented a check that stops the model if this is not the case. That's something we can discuss and maybe we find a better solution.
2. the differences in `obcs_ctrl` (I only found reduction of 16 to 14 in mean of (ADJ) U) come up in 
[d1a0161](https://github.com/MITgcm/MITgcm/pull/723/commits/d1a0161a0621567d537b5cf172c45b6c04cfc49e), where I store `cDrag` in 2 places (+2 if shelfice is used).  This avoids recomputing `cDrag`, which can explain the small difference. For `verification_other/global_oce_cs32` (and probably in `global_oce_llc`), I need more time to figure out the differences.

@owang01 I added a comment about removing the checkpointing code in the PR description (and tag-index suggestion)

jm-c commented:
@mjlosch I reverted the last commit to avoid conflicts when merging master in (since PR #760 has now been merged in).

jm-c commented on model/src/solve_pentadiagonal.F:
```diff
@@ -65,6 +65,9 @@ SUBROUTINE SOLVE_PENTADIAGONAL(
 
+C/*   This flag combination is only false if INCLUDE_IMPLVERTADV_CODE is
+C     not defined while ALLOW_AUTODIFF is. */
+#if !(!defined INCLUDE_IMPLVERTADV_CODE && (defined ALLOW_AUTODIFF))
```
The double negative here could easily be avoided, with simply:
"#if ( defined INCLUDE_IMPLVERTADV_CODE || !(defined ALLOW_AUTODIFF) )"

jm-c commented:
@mjlosch Regarding the changes related to GAD_ALLOW_TS_SOM_ADV which now needs to be defined to use SOM advection for ptracers: 

I looked to the archive, and it seems that this PR changes is reverting the modifications from May 09, 2008 (was under CVS, so likely not a single Git commit):
Before May 09, 2008, a single CPP option "GAD_ALLOW_SOM_ADVECT" was used a) to include the som_T & som_S common block /GAD_SOM_VARS_R/ ; and b) to compile the SOM advection S/R.
After May 09, 2008, the CPP option was renamed "GAD_ALLOW_TS_SOM_ADV" and only applies to the T & S SOM common block.

We could revisit this 2008 decision, but first I would like to know if/how the changes here are needed for the adjoint. Currently it's not obvious to me.

mjlosch commented:
@jm-c the changes of [1ee65c6](https://github.com/MITgcm/MITgcm/pull/723/commits/1ee65c635996bb6b269ba41ce8704482f89c19fa) are not strictly needed for the AD code, but they suppress many "not necessary" warning for cases when SOM is not complied. I can replace all
`#ifdef ALLOW_AUTODIFF_TAMC`
by
`#if ( defined ALLOW_AUTODIFF_TAMC && defined GAD_ALLOW_TS_SOM_ADV )`
in the three routines. Then forward code will not be affected but we would still need this flag defined for ptracers+AD. I guess that makes more sense. And I'll try it out.

mjlosch commented:
@jm-c I agree, that's probably nicer, because the current solution again implies a restriction (GAD_ALLOW_TS_SOM_ADV needs to be defined if you want to use AD som-code with passive tracers). I think I would like to change this again according to your suggestion, because it reduces the number of stupid cpp-flags. 

What do we do about the check in `ptracers_check.F`? It can probably be simplified/reduced to
```
#ifndef PTRACERS_ALLOW_DYN_STATE
       IF ( PTRACERS_SOM_Advection(iTracer) ) THEN
        WRITE(msgBuf,'(2A,I4,A)') 'PTRACERS_CHECK:',
     &    ' PTRACERS_SOM_Advection(', iTracer, ') = T, but'
        CALL PRINT_ERROR( msgBuf, myThid )
        WRITE(msgBuf,'(3A)') 'PTRACERS_CHECK:',
     &    ' PTRACERS_ALLOW_DYN_STATE is not defined',
     &    ' in PTRACERS_OPTIONS.h.'
        CALL PRINT_ERROR( msgBuf, myThid )
        errCount = errCount + 1
       ENDIF
#endif
```

Do you agree?

jm-c commented:
@mjlosch 
> Do you agree?

Yes

jm-c commented:
@mjlosch Are done with changes in this PR ? Once you are, will run full set of Adjoint tests to just double check.

mjlosch commented:
@jm-c I played a little bit with the `DISABLE_SIGMA_CODE` flag for which I have a solution now (although it is not clear to me, if this will ever be needed, just to avoid the recomputations if the flag is accidentally not defined), but I don't think that this belongs here, so I'll stop. I am done, please do your AD tests.

jm-c commented:
@mjlosch yes, no need to spend time with DISABLE_SIGMA_CODE now (I don't think this has eever been tested anyway).
All changes look good to me. I am just waiting for the `verification_other` AD tests to finish, both with new "cDrag" store and without, just to double check.

jm-c commented:
Got all the AD tests back, all good, minor differences in AD monitor only in `obcs_ctrl` and from `verification_other` in cs32 and llc90, and can confirm that these are caused by `cDrag` storage and storage in `calc_phi_hyd.F`.
Will merge this PR soon (later today or tomorrow).
